scripts/mbedtls_dev/bignum_common.py

```diff
@@ -78,22 +82,106 @@ class OperationCommon:
+            in the generated test cases. "variable" passes them as they are
+            defined in the python source. "arch_split" pads the values with
+            zeroes depending on the architecture/limb size. If this is set,
+            test cases are generated for all architectures.
+        arity: the number of operands for the operation. Currently supported
+            values are 1 and 2.
     """
     symbol = ""
-    input_values = [] # type: List[str]
-    input_cases = [] # type: List[Tuple[str, str]]
-    unique_combinations_only = True
-
-    def __init__(self, val_a: str, val_b: str) -> None:
-        self.arg_a = val_a
-        self.arg_b = val_b
+    input_values = INPUTS_DEFAULT # type: List[str]
+    input_cases = [] # type: List[Any]
+    unique_combinations_only = False
+    input_styles = ["variable", "fixed", "arch_split"] # type: List[str]
+    input_style = "variable" # type: str
+    limb_sizes = [32, 64] # type: List[int]
+    arities = [1, 2]
+    arity = 2
+
+    def __init__(self, val_a: str, val_b: str = "0", bits_in_limb: int = 32) -> None:
+        self.val_a = val_a
+        self.val_b = val_b
+        # Setting the int versions here as opposed to making them @properties
+        # provides earlier/more robust input validation.
         self.int_a = hex_to_int(val_a)
         self.int_b = hex_to_int(val_b)
+        if bits_in_limb not in self.limb_sizes:
+            raise ValueError("Invalid number of bits in limb!")
+        if self.input_style == "arch_split":
+            self.dependencies = ["MBEDTLS_HAVE_INT{:d}".format(bits_in_limb)]
+        self.bits_in_limb = bits_in_limb
+
+    @property
+    def boundary(self) -> int:
+        if self.arity == 1:
+            return self.int_a
+        elif self.arity == 2:
+            return max(self.int_a, self.int_b)
+        raise ValueError("Unsupported number of operands!")
+
+    @property
+    def limb_boundary(self) -> int:
+        return bound_mpi(self.boundary, self.bits_in_limb)
+
+    @property
+    def limbs(self) -> int:
+        return limbs_mpi(self.boundary, self.bits_in_limb)
+
+    @property
+    def hex_digits(self) -> int:
+        return 2 * (self.limbs * self.bits_in_limb // 8)
+
+    def format_arg(self, val) -> str:
+        if self.input_style not in self.input_styles:
+            raise ValueError("Unknown input style!")
+        if self.input_style == "variable":
+            return val
+        else:
+            return val.zfill(self.hex_digits)
+
+    def format_result(self, res) -> str:
+        res_str = '{:x}'.format(res)
+        return quote_str(self.format_arg(res_str))
+
+    @property
+    def arg_a(self) -> str:
+        return self.format_arg(self.val_a)
+
+    @property
+    def arg_b(self) -> str:
+        if self.arity == 1:
+            raise AttributeError("Operation is unary and doesn't have arg_b!")
+        return self.format_arg(self.val_b)
 
     def arguments(self) -> List[str]:
```

Comment:
This method is tricky and very critical since it is invoked in class creation. We are assuming certain things that we need to make clear.

* Result is expected to return a list.
* Result is expecting to implement its own quote_str or other formatting if needed

I am also wondering if it should be smarter, and apply some context with regards to auto-formatting the output. The less the children have to override this method, the less likely we are fall to code duplication.

I cannot see a way for the current implementation to handle things like `data_t`, int, and enum types eg `MBEDTLS_MPI_MOD_REP_MONTGOMERY` or `MBEDTLS_MPI_MOD_EXT_REP_BE`

The latter can be achieved by adding a arg_map class member which is a dictionary mapping special values to things 

~~~~~~
self.arg_map = {"99": "MBEDTLS_MPI_MOD_REP_MONTGOMERY", "199":"MBEDTLS_MPI_MOD_REP_OPT_RED"}                                           


def arguments(self) -> List[str]:
    args = [quote_str(self.arg_a)]
    ....
    .....
    
    args = [arg_map[n] if n in arg_map.keys() else n for n in args]
~~~~~~

But it feels that our data should have a context label to signal to the argument formatter on how to contert to the desired format.

Reply:
Indeed that looks complicated. I don't think we should make our existing classes more complicated/universal. Maybe we could add a child to `OperationCommon` that satisfies the needs of Input/Oputput/Conversion operations.

Either way, I would like for us to address this in a different PR, so that we can bring this PR in earlier.